### ISSUE TITLE
Report overall fenestration product U-factor, SHGC, and VT

### ISSUE THREAD

**Issue by rd2** [NREL/OpenStudio#4379]
Enhancement Request
--------------
Provide SDK users with functionality to access overall fenestration U-factor, SHGC, and VT ... similar to this (open) _new feature_ EnergyPlus [request](https://github.com/NREL/EnergyPlus/issues/6530)).

## Detailed Description
When linking fenestrated subsurfaces (e.g. GlassDoor, FixedWindow) with frame & divider (F&D) objects, there is no obvious solution for users to retrieve the overall fenestration U-factor (as well as SHGC & VT), similar to NFRC calculations. A given building design may have e.g. 3 fenestrated subsurface multilayered constructions, and 17 unique window sizes. The generic nature of F&D objects (i.e. a single F&D object may be reused for multiple fenestrated subsurfaces with unique geometries) offers OpenStudio/EnergyPlus users a very useful (automated) solution to generate a wide range of subsurface-specific properties (great). But in the end, users cannot retrieve the resulting NRFC-like aggregate metrics like overall U-factors. This is becoming an issue for pre-simulation input validation (and/or post-simulation output validation) for code compliance purposes in cold climates e.g., recent codes, upcoming revisions (currently in development).

## Possible Implementation
2 options:

1. (preferred) Implement the NFRC/LBNL Window calculations in the OpenStudio SDK to generate subsurface-specific U-factors (standard winter conditions), SHGC (standard summer conditions) ... and VT (less of a priority). A first prototype could be an OpenStudio Measure (or a common gem).
2. (quick fix) If users have F&D inputs (LBNL [Window](https://windows.lbl.gov/tools/window/software-download)) for a given fenestration product, they likely already have on hand NFRC U-factors for standard winter conditions. Allow SDK users to set a fenestrated subsurface [uFactor](https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-3.2.1-doc/model/html/classopenstudio_1_1model_1_1_planar_surface.html#a9404310b4e6e9cc5ebb5fe948be29f43) as a placeholder for pre-simulation input validation (currently [not allowed](https://github.com/NREL/OpenStudio/blob/a60c4629d3de20551295e6c131aec25a9ead22c2/src/model/SubSurface.cpp#L457)). The would not provide the desired flexibility described above, but better than nothing.

Both options would have no bearing on EnergyPlus inputs and/or simulation results.

**Comment by macumber** [comment]
This would be a good potential first example of linking OpenStudio to either the EnergyPlus C libraries or linking to Windows-CalcEngine directly.  The calculations for window properties are too detailed to reproduce IMHO, it would be better to just link to the same libraries that EnergyPlus uses directly.

**Comment by shorowit** [comment]
@rd2 Is your "quick fix" simply the desire to tag the U-factor on a subsurface? If so, you can use the `additionalProperties` object for that.

```
subsurface.additionalProperties.setFeature("uFactor", Float(ufactor))
puts subsurface.additionalProperties.getFeatureAsDouble("uFactor").get
```

**Comment by rd2** [comment]
As per @shorowit 's suggestion, option 2 (the _quick fix_) is easy to set up with AdditionalProperties. Implemented [here](https://github.com/rd2/tbd/blob/e92a7b9561074340321b0c10827418603bc8ed7f/lib/framedivider.rb#L212), and tested [here](https://github.com/rd2/tbd/blob/e92a7b9561074340321b0c10827418603bc8ed7f/spec/tbd_tests_spec.rb#L4876). Thanks!

**Comment by jmarrec** [comment]
I'm not sure I'm following the intent correctly. But it seems you can just run a single day of simulation or whatnot and get it from E+ SQL file. Surface::uFactor() can be used to do that SQL query for you.

**Comment by rd2** [comment]
It doesn't look like E+ is [there yet](https://github.com/NREL/EnergyPlus/pull/8740) (soonish? 9.6?).

**Comment by jmarrec** [comment]
Ok, for assembly, E+ has yet to report that. But the point remains the same. If E+ can do that for you currently, then do a run and retrieve the value. Once https://github.com/NREL/EnergyPlus/pull/8740/ is merged in we can add a helper function to query the Sql, sure.

-----

Linking to the E+ C API is a nice idea, but the C API is currently very shallow and definitely doesn't include this type of functions. It exposes only functions for EMS-like features and some _simple_ others like Psychrometric ones (where you pass in a couple of doubles and that's it, eg https://github.com/NREL/EnergyPlus/blob/55c844eb1883650f490889137692aab8b58fd103/src/EnergyPlus/api/func.cc#L163-L170)

These specific calculations require the internal state of E+ to be initialized by (a ton of) routines in a specific order so that the data structures are populated accordingly for the calculation to succeed. These routines also often don't just do the specific bit you are interested in but more than that, so you must initialize a bunch of other stuff ("globals") if you want to run a specific routine and not have it crash. Trust me on that, this is why writing unit tests for E+ is really really painful.

Anyways, we can't really just pass in a few numbers and get something out of it, so to me it looks like a big no go right now.

**Comment by macumber** [comment]
I added https://github.com/NREL/OpenStudio/issues/4403 to request the Sql helpers after NREL/EnergyPlus#8740 is merged

**Comment by jmarrec** [comment]
Can I close this one to be superseded by #4403 then?

**Comment by macumber** [comment]
It sounds like these calculations would not be a good fit for the E+ API so it is probably best to just link OS with https://github.com/LBNL-ETA/Windows-CalcEngine directly

**Comment by jmarrec** [comment]
Note that it'd require building Windows-CalcEngine as part of the build process. libWindows-CalcEngine.a is 5MB is release and 1484 MB in Debug.

I personally think that this is more trouble that the benefits it brings, but I'll defer to @tijcolem on this.

**Comment by tijcolem** [comment]
This just popped up during iteration meeting.  I am fine keeping this open as longer term feature request. I don't know what's involved in using the Window-calcengine. We'd need to make sure this would be supported cross-platform via conan, but it sounds like it would.
